webapp/packages/plugin-data-viewer/src/ValuePanelPresentation/ImageValue/ImageValuePresentation.tsx

```diff
@@ -36,58 +41,171 @@ const styles = css`
+        </tools>
+      )}
+      {onToggleStretch && (
+        <tools>
+          <tools-action
+            as='button'
+            title={translate('data_viewer_presentation_value_image_fit')}
+            disabled={stretch}
+            onClick={onToggleStretch}
+          >
+            <IconOrImage icon='img-fit-size' />
+          </tools-action>
+          <tools-action
+            as='button'
+            title={translate('data_viewer_presentation_value_image_original_size')}
+            disabled={!stretch}
+            onClick={onToggleStretch}
+          >
+            <IconOrImage icon='img-original-size' />
+          </tools-action>
+        </tools>
+      )}
+    </tools-container>
+  );
+});
+
 export const ImageValuePresentation: TabContainerPanelComponent<IDataValuePanelProps<any, IDatabaseResultSet>> = observer(function ImageValuePresentation({
   model,
   resultIndex,
 }) {
   const translate = useTranslate();
-  const [stretch, setStretch] = useState(false);
-  const selection = model.source.getAction(resultIndex, ResultSetSelectAction);
+  const notificationService = useService(NotificationService);
+  const style = useStyles(styles);
 
-  const focusCell = selection.getFocusedElement();
+  const state = useObservableRef(() => ({
+    get selectedCell() {
+      const selection = this.model.source.getAction(this.resultIndex, ResultSetSelectAction);
+      const focusCell = selection.getFocusedElement();
 
-  let src: string | undefined;
+      return selection.elements[0] || focusCell;
+    },
+    get cellValue() {
+      const view = this.model.source.getAction(this.resultIndex, ResultSetViewAction);
+      const cellValue = view.getCellValue(this.selectedCell);
 
-  if (selection.elements.length > 0 || focusCell) {
-    const view = model.source.getAction(resultIndex, ResultSetViewAction);
-    const firstSelectedCell = selection.elements[0] || focusCell;
+      return cellValue;
+    },
+    get src() {
+      if (this.savedSrc) {
+        return this.savedSrc;
+      }
 
-    const cellValue = view.getCellValue(firstSelectedCell);
+      if (isResultSetContentValue(this.cellValue) && this.cellValue.binary) {
+        return `data:${getMIME(this.cellValue.binary)};base64,${this.cellValue.binary}`;
+      } else if (typeof this.cellValue === 'string' && isValidUrl(this.cellValue) && isImageFormat(this.cellValue)) {
+        return this.cellValue;
+      }
 
-    if (isResultSetContentValue(cellValue) && cellValue.binary) {
-      src = `data:${getMIME(cellValue.binary)};base64,${cellValue.binary}`;
-    } else if (typeof cellValue === 'string' && isValidUrl(cellValue) && isImageFormat(cellValue)) {
-      src = cellValue;
-    }
+      return '';
+    },
+    get savedSrc() {
+      return this.model.source.dataManager.retrieveFileDataUrlFromCache(this.selectedCell, this.resultIndex);
+    },
+    get canSave() {
+      return this.model.source.dataManager.canDownload(this.selectedCell, this.resultIndex);
```

Comment:
it's probably always true for images (the exception is when we show a placeholder)